README.md
<div align="center">

<!-- TODO: on 1.0.0 release, delete this section -->

| :warning: WARNING: this project is considered to be in BETA (pre-`1.x.x`), `1.0.0` release imminent! |
| ---------------------------------------------------------------------------------------------------- |

<!-- END -->

![SvelteKit adapter Firebase social preview](assets/github-preview-svelte-adapter-firebase.png)

# svelte-adapter-firebase

[![GitHub Release](https://img.shields.io/github/release/jthegedus/svelte-adapter-firebase.svg?color=green)](https://github.com/jthegedus/svelte-adapter-firebase/releases)
[![npm](https://img.shields.io/npm/v/svelte-adapter-firebase?color=green)](https://www.npmjs.com/package/svelte-adapter-firebase)
[![Tests](https://github.com/jthegedus/svelte-adapter-firebase/actions/workflows/test.yaml/badge.svg)](https://github.com/jthegedus/svelte-adapter-firebase/actions/workflows/test.yaml)
[![CodeQL](https://github.com/jthegedus/svelte-adapter-firebase/actions/workflows/codeql-analysis.yaml/badge.svg)](https://github.com/jthegedus/svelte-adapter-firebase/actions/workflows/codeql-analysis.yaml)

[Firebase](https://firebase.google.com/) adapter for
[SvelteKit](https://github.com/sveltejs/kit).

Utilise the Firebase Hosting CDN with dynamic content served by SvelteKit on
Cloud Functions!

:heavy_check_mark: SSR on
[Cloud Functions](https://firebase.google.com/docs/hosting/functions)</br>
:heavy_check_mark: Integrates with existing Cloud Functions!</br>
:heavy_check_mark: Local production testing with
[Firebase Emulator](https://firebase.google.com/docs/emulator-suite)</br>
:heavy_check_mark: Mitigate cold-starts with
[minInstances](https://firebase.google.com/docs/functions/manage-functions#min-max-instances)</br>
:heavy_check_mark:
[Multiple Hosting Sites](https://firebase.google.com/docs/hosting/multisites#add_additional_sites)</br>

</div>

## Contents

- [Contents](#contents)
- [Setup](#setup)
  - [Beta Adapter Version Compatibility](#beta-adapter-version-compatibility)
- [Configuration Overview](#configuration-overview)
- [Details](#details)
  - [`firebase.json` Configurations](#firebasejson-configurations)
  - [Adapter Configurations](#adapter-configurations)
- [How it works](#how-it-works)
- [Firebase Emulator local Testing](#firebase-emulator-local-testing)
- [Deployment](#deployment)
- [Caveats](#caveats)
- [Non-goals](#non-goals)
- [FAQ](#faq)
- [Contributing](#contributing)
  - [external contributions](#external-contributions)

## Setup

The adapter reads `firebase.json` to determine output dirs for Server scripts &
Static assets, without this file the adapter cannot know how your Firebase app
is configured. Hosting & Cloud Functions are required.

In your standard SvelteKit project:

- `npm install --save-dev svelte-adapter-firebase`
- add adapter to `svelte.config.js`:

```diff
+import firebase from "svelte-adapter-firebase";

/** @type {import('@sveltejs/kit').Config} */
export default {
  kit: {
+   adapter: firebase(),
  },
};
```

- Setup `firebase.json` with `firebase init`.
- `npm run build`. **Read and repeat, the output is meant as a guide. IE: after
  your first build, you need to manually add the produced Cloud Function in
  `index.js`**

<!-- TODO: on 1.0.0 release, delete this section -->

### Beta Adapter Version Compatibility

SvelteKit is still in Beta and the Adapter API is in flux, this can result in
the Adapter and SvelteKit becoming incompatible. Here is a compatibility table:

| Adapter Version | SvelteKit Version    |
| --------------- | -------------------- |
| `0.15.0`        | `1.1.1`              |
| `0.14.5`        | `1.0.0-next.587`     |
| `0.14.4`        | `1.0.0-next.443`     |
| `0.14.3`        | `1.0.0-next.443`     |
| `0.14.2`        | `1.0.0-next.405`     |
| `0.14.0`        | `1.0.0-next.330`     |
| `0.13.1`        | `1.0.0-next.180`     |
| `0.13.0`        | `1.0.0-next.168`     |
| `0.12.x`        | `1.0.0-next.165`     |
| `0.11.x`        | `1.0.0-next.155`     |
| `NA`            | `1.0.0-next.152-154` |
| `0.10.x`        | `1.0.0-next.132`     |
| `0.9.1`         | `1.0.0-next.122`     |
| `0.9.0`         | `1.0.0-next.120`     |
| `0.8.x`         | `1.0.0-next.111`     |
| `NA`            | `1.0.0-next.109-110` |
| `0.7.x`         | `1.0.0-next.107`     |
| `0.6.x`         | `1.0.0-next.103`     |
| `0.5.x`         | `1.0.0-next.54`      |
| `0.4.x`         | `1.0.0-next.46`      |
| `0.3.x`         | `1.0.0-next.27`      |

**Note**: only the versions listed have been tested together, if others happen
to work, it is just coincidence. This is beta software after all.

<!-- END -->

## Configuration Overview

Adapter options:

- `esbuildBuildOptions`
  - function to return an `esbuild.BuildOptions` object
  - default: see `defaultOptions` object in [`src/index.js`](./src/index.js)
- `firebaseJsonPath`
  - path to your `firebase.json` file, **relative** from where `svelte build` is
    called
  - default: `./firebase.json`
- `target`
  - required when `firebase.json:hosting` is an array (contains many site
    configurations)
  - default: `undefined`
- `sourceRewriteMatch`
  - used to lookup the rewrite config to determine whether to output SSR code
    for Cloud Functions. See
    [Firebase Rewrite configuration docs](https://firebase.google.com/docs/hosting/full-config#rewrite-functions).
  - default: `**`

Adapter output:

- static assets (images, CSS, Client-side JavaScript) of your SvelteKit app
  output to the directory defined by `firebase.json:hosting.public`
- server assets (SSR JavaScript) output alongside your Cloud Functions defined
  by `firebase.json:functions.source`

## Details

[Setup](#setup) outlines the steps most commonly used with a single SvelteKit
app. Here we go into the details of each configuration and how it interacts with
the `firebase.json` config.

The 3 step process is:

1. select Hosting config from `firebase.json`. If more than one site present in
   config, match `svelte.config.js:target` field with either
   `firebase.json:hosting[].site` or `.target` fields.
2. output static assets to the directory in the `public` field
3. identify the rewrite rule for SSR to determine Cloud Function output. The
   rewrite rule is determined by a lookup of the `rewrites.source` against
   `sourceRewriteMatch`

### `firebase.json` Configurations

Due to the relaxed rules of `firebase.json` we can have many valid configs. At a
minimum, one or more Hosting sites is required with an associated Functions
config if a Cloud Function rewrite is used. These are the combintations:

<details>
<summary>single Hosting site with Cloud Function rewrite</summary>

```json
{
  "hosting": {
    "public": "<someDir>",
    "rewrites": [
      {
        "source": "**",
        "function": "<functionName>"
      }
    ]
  },
  "functions": {
    "source": "<anotherDir>"
  }
}
```

</details>

<details>
<summary>multiple Hosting site with Cloud Function rewrite</summary>

```json
{
  "hosting": [
    {
      "site": "blog",
      "public": "<someDir>",
      "rewrites": [
        {
          "source": "**",
          "function": "<functionName>"
        }
      ]
    },
    {
      // another site config
    }
  ],
  "functions": {
    "source": "<anotherDir>"
  }
}
```

To correctly lookup the `blog` site, `target` will need to be set in
`svelte.config.js`:

```js
import firebase from "svelte-adapter-firebase";

/** @type {import('@sveltejs/kit').Config} */
export default {
  kit: {
    adapter: firebase({ target: "blog" }),
  },
};
```

</details>

### Adapter Configurations

Detailed examples of the adapter configuration options.

All options:

```js
import firebase from "svelte-adapter-firebase";

/** @type {import('@sveltejs/kit').Config} */
export default {
  kit: {
    adapter: firebase({
      esbuildBuildOptions: (defaultOptions: BuildOptions) => Promise<BuildOptions> | BuildOptions,
      firebaseJsonPath: "",
      target: "svelte-func-single-site",
      sourceRewriteMatch: "",
    }),
  },
};
```

<details>
<summary><code>esbuildBuildOptions</code></summary>

As an escape hatch, you may optionally specify a function which will receive the
final esbuild options generated by this adapter and returns a modified esbuild
configuration. The result of this function will be passed as-is to esbuild. The
function can be async.

For example, you may wish to add `plugins`, or configure the
[`format`](https://esbuild.github.io/api/#format) to bundle to ESM (defaults to
CJS):

```js
import firebase from "svelte-adapter-firebase";

/** @type {import('@sveltejs/kit').Config} */
export default {
  kit: {
    adapter: firebase({
      target: "svelte-func-single-site",
      esbuildBuildOptions(defaultOptions) {
        return {
          ...defaultOptions,
          target: "esm"
          plugins: [],
        };
      },
    })
  },
};
```

The default options for this version are as follows:

```js
{
	entryPoints: ['.svelte-kit/firebase/handler.js'],
	outfile: `pathToOutputDir/index.js`,
	bundle: true,
	inject: ['pathTo/shims.js'],
	platform: 'node',
  target: `node${functionRuntimeVersion}`
}
```

where esbuild `target` is computed from the Node.js runtime version defined for
your Cloud Functions.

</details>

<details>
<summary><code>firebaseJsonPath</code></summary>

If the `firebase.json` file is not in the directory you run `svelte build`, then
you can set a relative path in `svelte.config.js`:

```
.gitignore
firebase.json
app/                    <-- svelte build run in this dir
	package.json
	svelte.config.js
	src/
anotherApp/
	index.html
	index.css
functions/
	package.json
	index.js
```

```js
import firebase from "svelte-adapter-firebase";

/** @type {import('@sveltejs/kit').Config} */
export default {
  kit: {
    adapter: firebase({
      target: "svelte-func-single-site",
      firebaseJsonPath: "../firebase.json",
    }),
  },
};
```

</details>

<details>
<summary><code>target</code></summary>

If `firebase.json:hosting` is an array of sites, then each hosting config must
list a `site` or `target` field that matches the adatper's `target` option. For
example:

```json
// firebase.json
{
  "hosting": [
    {
      "site": "blog",
      // or
      // "target": "blog",
      "public": "<someDir>",
      "rewrites": [
        {
          "source": "**",
          "run": {
            "serviceId": "<cloudRunServiceId>"
          }
        }
      ]
    },
    {
      "site": "adminPanel",
      // or
      // "target": "adminPanel",
      "public": "<anotherDir>"
    }
  ]
}
```

```js
import firebase from "svelte-adapter-firebase";

/** @type {import('@sveltejs/kit').Config} */
export default {
  kit: {
    adapter: firebase({ target: "blog" }),
    target: "#svelte",
  },
};
```

The Firebase config & adapter config match (`firebase.json:hosting[0].site` ===
adapter `target`), so therefore we know which Firebase Hosting site you want to
build the SvelteKit site for.

</details>

<details>
<summary><code>sourceRewriteMatch</code></summary>

If the rewrite `source` pattern is not `**`, then `svelte.config.js`
`sourceRewriteMatch` will need to be set to match your desired rewrite rule. For
example:

```json
// firebase.json
{
  "hosting": {
    "public": "<someDir>",
    "rewrites": [
      {
        "source": "/blog/**",
        "run": {
          "serviceId": "<cloudRunServiceId>"
        }
      }
    ]
  }
}
```

```js
import firebase from "svelte-adapter-firebase";

/** @type {import('@sveltejs/kit').Config} */
export default {
  kit: {
    adapter: firebase({ sourceRewriteMatch: "/blog/**" }),
    target: "#svelte",
  },
};
```

</details>

## How it works

Given

- the following `firebase.json` configuration
- a standard SvelteKit app structure
- the default `svelte-adapter-firebase` config

```json
// firebase.json
{
  "hosting": {
    "public": "myApp",
    "rewrites": [
      {
        "source": "**",
        "function": "ssrServer"
      }
    ],
    "predeploy": ["npm run build"]
  },
  "functions": {
    "source": "functions"
  }
}
```

the following Server & Static assets dirs are created:

```
firebase.json ("public": "myApp")
package.json
svelte.config.js
src/
	app.html
	routes/
		index.svelte
functions/
	package.json ("main": "index.js")
	index.js
	sveltekit/		<-- Server Assets dir (code to be imported to you Cloud Function)
myApp/				<-- Static Assets to go to Firebase Hosting CDN
```

- `firebase.json:functions.source` dir is used to find `functions/package.json`
  whose `main` field is used to find the Cloud Function build dir. This is used
  as the server asset output dir.

<details>
  <summary>TypeScript Cloud Functions</summary>

Because we use the above method to determine the output dir, the server assets
are output to the correct place when using TypeScript.

```
firebase.json ("public": "myApp")
package.json
svelte.config.js
src/
  app.html
  routes/
    index.svelte
functions/
  package.json ("main": "lib/index.js")
  index.ts
  lib/
    index.js
    sveltekit/	<-- Server assets output to functions/lib
myApp/				<-- Static assets to go to Firebase Hosting CDN
```

</details>

<details>
<summary>Output with Multiple Sites</summary>

In a multi-site setup, the `site` or `target` field from hosting config in
`firebase.json` is used as the server output dir:

```
firebase.json ("site": "myCoolSite","public": "myApp")
package.json
svelte.config.js
src/
	app.html
	routes/
		index.svelte
functions/
	package.json
	index.js
	myCoolSite/		<-- Server assets
myApp/				<-- Static assets to go to Firebase Hosting CDN
```

</details>

The final piece is to write the actual Cloud Function source code to reference
the output server assets. The code is printed during `svelte build` and should
be placed in your `index.js` or `index.ts` manually.

This is a flexible solution that allows integrating with other Cloud Functions
in your project. You can edit the provided code as you see fit. The
import/require of the generated code will not change unless you change the
`firebase.json:hosting.site` or `package.json:main` fields, so you shouldn't
need to update this code after adding it.

## Firebase Emulator local Testing

Test your production build locally before pushing to git or deploying!

- build your app: `svelte-kit build`
- install Function dependencies: `pnpm install --prefix functions`
- start the emulator: `firebase emulators:start`

## Deployment

`firebase deploy` :tada:

## Caveats

- Using `firebase.json:hosting[].site` is preferred to
  `firebase.json:hosting[].target` as
  [Firebase Deploy Targets](https://firebase.google.com/docs/cli/targets) only
  supports Hosting, Storage & Databases and not Functions. This means you can
  use Deploy targets `target` field to identify your site for the adapter to
  **build**, but you **CANNOT use Deploy Targes when deploying** as you need to
  deploy the Hosting _& Functions_ at the same time for this solution to work as
  expected.
- [Firebase Hosting Preview Channels](https://firebase.google.com/docs/hosting/test-preview-deploy)
  currently lacks first-party support for SSR applications. This adapter doesn't
  attempt to remedy this issue and doesn't produce a different SSR Function for
  preview channel deployments.
- :warning: Cloud Function rewrites only support **us-central1**, other regions
  will error. The official warning about this can be found in
  [these docs](https://firebase.google.com/docs/hosting/functions).

## Non-goals

> Write Cloud Function code directly into `.js` file instead of printing in
> console.

Firebase Cloud Functions have a long history of people configuring their index
files completely differently, some even generating them from directories.
Accommodating these would be a headache. Instead, all we look for is a match
against this string, `${name} =`, where `name` is your Cloud Functions name. We
may make this configurable to a specific file in future.

Additionally, this allows for users to customise their Firebase Cloud Function
API like `runWith()` options for memory/CPU, min/max Instances and
VPC/Ingress/Egress configuration settings, without complex support for options
in the adapter. This keeps the Function config where it should, close to the
userland code.

> Handle the deployment of the app to Firebase.

Firebase apps consist of many different services with the CLI providing optional
deployments. We do not want to dictate full deployments with your frontend nor
perform partial deployments if it does not fit your app. The only option then is
to leave it to you :tada:

## FAQ

> Why is the Cloud Function code output to the terminal for me to add manually
> instead of being written to `functions/index.js`?

See [non-goals](#non-goals) _Write Cloud Function code directly into `.js` file
instead of printing in console._

> Firebase libs in SvelteKit

As recommended in the [SvelteKit FAQ](https://kit.svelte.dev/faq), please use
[Firebase JS SDK v9](https://firebase.google.com/docs/web/learn-more#modular-version)
as the older version of the SDK has issues and a larger bundle size.

> Cold Starts

Since the purpose of using this adapter is to leverage the Firebase Hosting CDN,
you should consider improving the user experience with targetted caching/TTLs.

If cold start are still an issue for your application, Cloud Functions has
support for `minInstances` which will keep `X` number of instances warm. From
the docs:

> A minimum number of instances kept running incur billing costs at idle rates.
> Typically, to keep one idle function instance warm costs less than $6.00 a
> month. The Firebase CLI provides a cost estimate at deployment time for
> functions with reserved minimum instances. Refer to Cloud Functions Pricing to
> calculate costs.
>
> -[Firebase docs](https://firebase.google.com/docs/functions/manage-functions#min-max-instances)

To implement this, configure your
[`runWith`](https://github.com/firebase/firebase-functions/blob/d46ec6191e61f560f3f21f13333e0f3285d3de90/src/function-configuration.ts#L101)
options like so:

```diff
const myRuntimeOptions = {
	memory: "1GB",
+	minInstances: 1,
}
exports.myFunc = functions.runWith(myRuntimeOptions).https.onRequest(async (request, response) => {
	...
});
```

Note: this is still single concurrency (if an instance does not exist to handle
a request when it hits the backend a new Function instance is created). Watch
this space!

<!-- TODO: on 1.0.0 release, delete this section -->

- `1.0.0` will not be published until the SvelteKit Adapter API is declared
  stable and SvelteKit is released for general use.

<!-- END -->

## Contributing

[Contributions of any kind welcome, just follow the
guidelines](CONTRIBUTING.md)!

Short version:

```
git clone https://github.com/jthegedus/svelte-adapter-firebase.git
asdf install
pnpm i
```

See [asdf](https://asdf-vm.com) to install set it up.

### external contributions

While building this adapter some issues were found with upstream components,
these are captured here should someone wish to contribute them:

- Cloud Function validation code linked in `utils.js` is from two different
  sources which indicates that it is being validated by `firebase-tools` in two
  separate places. PR a fix there.


CHANGELOG.md
# Changelog

## [0.15.0](https://www.github.com/jthegedus/svelte-adapter-firebase/compare/v0.14.4...v0.15.0) (2023-01-17)


### ⚠ BREAKING CHANGES

* sync with kit@1.1.1 based on (PRs #194 #200) (#203)

### Bug Fixes

* **#192:** replace getStaticDirectory with kit.files.assets ([#193](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/193)) ([a4b4daa](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/a4b4daac795cd93a4a06f5d218f69841810e686b))
* sync with kit@1.1.1 based on (PRs [#194](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/194) [#200](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/200)) ([#203](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/203)) ([352072d](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/352072d0f30ef76442fc143ccb4a446119a573ee))

### [0.14.4](https://www.github.com/jthegedus/svelte-adapter-firebase/compare/v0.14.3...v0.14.4) (2022-09-08)


### Bug Fixes

* cross-site POST requests ([#184](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/184)) ([e2d8081](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/e2d80812976feb5d16e34131462628647af59dc3))

### [0.14.3](https://www.github.com/jthegedus/svelte-adapter-firebase/compare/v0.14.2...v0.14.3) (2022-09-02)


### Bug Fixes

* **#180:** Prepend `Server.init` before calling `Server.respond` ([#181](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/181)) ([5c937c9](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/5c937c9ca6627025ecd19c889a9be3b50be2a266))

### [0.14.2](https://www.github.com/jthegedus/svelte-adapter-firebase/compare/v0.14.1...v0.14.2) (2022-08-13)


### Bug Fixes

* **#172:** Remove reference to obsoleted `writeStatic` ([#173](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/173)) ([9a990a0](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/9a990a00667b5f2197915d8b380593a5a05d306b))
* bump kit versions, update deps & fix xojs issues ([#178](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/178)) ([dde0a0e](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/dde0a0ea9802000df2b07c760357d8eac8a976fa))
* Pass rendered.headers to writeHead as object ([#176](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/176)) ([11d90cb](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/11d90cb6059585e9c95ca06d06e153ddf514f7b4))

### [0.14.1](https://www.github.com/jthegedus/svelte-adapter-firebase/compare/v0.14.0...v0.14.1) (2022-06-15)


### Bug Fixes

* **#169:** Replace `installFetch()` with `installPolyfills()` ([#170](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/170)) ([9c44f0c](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/9c44f0cbcab64a390ad4197f8a13f474761e5476))
* builds on windows ([#164](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/164)) ([0dd5005](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/0dd50055706e85d4fb1cff562ab73af42c93df33))

### [0.14.1](https://www.github.com/pham/svelte-adapter-firebase/compare/v0.14.0...v0.14.1) (2022-06-02)


### Bug Fixes

* **#169:** Replace @sveltejs/kit/install-fetch with @sveltejs/kit/node/polyfills ([cc00699](https://www.github.com/pham/svelte-adapter-firebase/commit/cc00699fdf968a6a771a644ea75cc25877ed9dc2))

## [0.14.0](https://www.github.com/jthegedus/svelte-adapter-firebase/compare/v0.13.1...v0.14.0) (2022-05-14)


### Features

* Support for SvelteKit Next 303 ([#159](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/159)) ([d10fcf6](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/d10fcf6b68b65f59363aa9d25210b0f9f2b63793))
* upgrade to sveltekit 329 ([#161](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/161)) ([bf7c2e9](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/bf7c2e90e61f7562993d465bc90b4ab1232dfaac))


### Bug Fixes

* reading 'log' error ([#157](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/157)) ([f855bfb](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/f855bfb7a18ed9186463e8a04af8e4ef886d3b6a))

## [0.14.0](https://www.github.com/jthegedus/svelte-adapter-firebase/compare/v0.13.1...v0.14.0) (2022-05-13)

### Bug Fixes

- Supports the latest sveltekit apis (tested up until sveltekit 330) ([[#159](https://github.com/jthegedus/svelte-adapter-firebase/issues/159))
- Thanks to nielsvandermolen and co3k for the patch submissions.

### [0.13.1](https://www.github.com/jthegedus/svelte-adapter-firebase/compare/v0.13.0...v0.13.1) (2021-10-10)

### Bug Fixes

- missed rename of hostingSite to target, bump kit for types ([#149](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/149)) ([cebf821](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/cebf8210c22291967b8ee2cbf3511736ddfbaef1))

## [0.13.0](https://www.github.com/jthegedus/svelte-adapter-firebase/compare/v0.12.0...v0.13.0) (2021-09-19)

### Features

- fb deploy targets. Rename opt `hostingSite` to `target` ([#144](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/144)) ([2ffe777](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/2ffe7774d41de9d4f88dc7d7935e5397e5999b44))

### Bug Fixes

- e2e tests with latest emulator ([#140](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/140)) ([5227bb5](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/5227bb5ced4576b530d0b9e6c38f91920e9863fe))
- pass null if no req.rawBody ([#143](https://www.github.com/jthegedus/svelte-adapter-firebase/issues/143)) ([15c9ca4](https://www.github.com/jthegedus/svelte-adapter-firebase/commit/15c9ca4bbd6783ab43614cb6b49692e4685372b8))

## [0.12.0](https://www.github.com/jthegedus/svelte-adapter-firebase/compare/v0.11.2...v0.12.0) (2021-09-16)

### Features


src/utils.js
import {existsSync, readFileSync} from 'fs';
import path from 'path';
import process from 'process';

/**
 * @typedef CloudRunRewriteConfig
 * @type {object} cloudrun
 * @property {undefined|'us-west1'} cloudrun.region
 * @property {undefined|string} cloudrun.serviceId
 */

/**
 * @typedef HostingRewriteConfig
 * @type {object} rewrite
 * @property {undefined|string} rewrite.function
 * @property {undefined|CloudRunRewriteConfig} rewrite.run
 * @property {undefined|string} rewrite.source
 */

/**
 * @typedef HostingConfig
 * @type {object} hosting
 * @property {undefined|string} hosting.public
 * @property {undefined|Array.<HostingRewriteConfig>} hosting.rewrites
 * @property {undefined|string} hosting.site
 * @property {undefined|string} hosting.target
 */

/**
 * @typedef FunctionsConfig
 * @type {object} functions
 * @property {undefined|'nodejs14'|'nodejs16'} functions.runtime
 * @property {undefined|string} functions.source
*/

/**
 * Firebase configuration from `firebase.json`. Typed with the types required by the adapter.
 *
 * @typedef FirebaseConfig
 * @type {object} config
 * @property {undefined|FunctionsConfig} functions
 * @property {undefined|HostingConfig|Array.<HostingConfig>} hosting
 */

/**
 *
 * @param {any} parameter
 * @returns {boolean} true if param is a string
 */
function isString(parameter) {
	return (typeof parameter === 'string' || parameter instanceof String);
}

/**
 * Extract Hosting Config from FirebaseConfig
 * @type {FirebaseConfig}
 * sourceRewriteMatch: string;
 * target: string|undefined;
 */
function extractHostingConfig(firebaseConfig, sourceRewriteMatch, target) {
	// Force "hosting" field to be array
	const firebaseHostingConfig = Array.isArray(firebaseConfig.hosting)
		? firebaseConfig.hosting
		: [{...firebaseConfig.hosting}];

	// Force "site" field to be included in "hosting" if more than 1 hosting site config
	if (firebaseHostingConfig.length > 1) {
		for (const item of firebaseHostingConfig) {
			if (!item.site && !item.target) {
				throw new Error('Error: Multiple "hosting" configurations found, each requires either a "site" field or "target" field, one does not. https://firebase.google.com/docs/hosting/multisites');
			}
		}
	}

	const hostingConfig = firebaseHostingConfig.length === 1
		? firebaseHostingConfig[0]
		: firebaseHostingConfig.find(item => (item.site === target && item.site !== undefined) || (item.target === target && item.target !== undefined));

	if (!hostingConfig) {
		if (!target) {
			throw new Error('Error: Multiple "hosting" configurations found, but no "target" specified in "svelte.config.js" adapter config. Provide one so we can match the config correctly.');
		}

		const hostingConfigValues = firebaseHostingConfig.map(item => (item.target ? {target: item.target} : {site: item.site}));
		throw new Error(`Error: Multiple "hosting" configurations found in "firebase.json" but not match found for ${target} specified in "svelte.config.js" adapter config. "hosting[].site" & "hosting[].target" values ${JSON.stringify(hostingConfigValues)}`);
	}

	if (!isString(hostingConfig?.public)) {
		throw new Error('Error: Required "hosting.public" field not found for hosting configuration.');
	}

	if (isString(hostingConfig?.public) && hostingConfig.public === '') {
		throw new Error('Error: Required "hosting.public" field is an empty string, a directory is required.');
	}

	if (!Array.isArray(hostingConfig?.rewrites)) {
		throw new TypeError(`Error: Required "hosting[].rewrites" field not found for matched hosting configuration. Specify your Cloud Function with rewrite rule matching "source":"${sourceRewriteMatch}"`);
	}

	return hostingConfig;
}

/**
 * Parse provided firebase.json to mat
 against SvelteKit config for target & Rewrite rule.
 *
 * @param {{
 * 	firebaseJsonPath: string
 * 	sourceRewriteMatch: string;
 * 	target: string|undefined;
 * }} param
 * @returns {{
 * 	functions: { name: string, source: string, runtime: string | undefined };
 * 	publicDir: string
 * }} Functions config with `public` dir
 */
function parseFirebaseConfiguration({target, sourceRewriteMatch, firebaseJsonPath}) {
	const firebaseJson = path.resolve(firebaseJsonPath);

	if (!existsSync(firebaseJson)) {
		throw new Error(`Error: The adapter requires a "firebase.json" file. "firebaseJsonPath:${firebaseJsonPath}" does not exist.`);
	}

	/**
	 * @type {FirebaseConfig}
	 */
	let firebaseConfig;
	try {
		firebaseConfig = JSON.parse(readFileSync(firebaseJson, 'utf8'));
	} catch (error) {
		throw new Error(`Error: failure while parsing ${firebaseJsonPath}. ${error.message}`);
	}

	if (!firebaseConfig?.hosting) {
		throw new Error('Error: "hosting" config missing from "firebase.json"');
	}

	const hostingConfig = extractHostingConfig(firebaseConfig, sourceRewriteMatch, target);

	const rewriteConfig = hostingConfig.rewrites.find(item => item.source === sourceRewriteMatch && (item.function || item.run));

	if (!rewriteConfig) {
		throw new Error(`Error: Required "hosting[].rewrites[]" does not contain a config with "source":"${sourceRewriteMatch}" and either "function":"<func_name>" or "run":{...} entries`);
	}

	if (rewriteConfig?.run && (!rewriteConfig.run.serviceId || !isString(rewriteConfig.run.serviceId))) {
		throw new Error('Error: Required "serviceId" field not found for Cloud Run rewrite rule in "firebase.json"');
	}

	if (rewriteConfig?.run && !validCloudRunServiceId(rewriteConfig.run.serviceId)) {
		throw new Error('Error: Cloud Run "serviceId" must use only lowercase alphanumeric characters and dashes cannot begin or end with a dash, and cannot be longer than 63 characters.');
	}

	if (rewriteConfig?.run && rewriteConfig?.run?.region && (rewriteConfig?.run?.region !== 'us-west1')) {
		throw new Error('Error: Cloud Run "region" is invalid, it should be "use-west1".');
	}

	if (rewriteConfig?.function && !validCloudFunctionName(rewriteConfig.function)) {
		throw new Error('Error: Cloud Function name must use only alphanumeric characters and underscores and cannot be longer than 63 characters');
	}

	// If function, ensure function root-level field is present
	if (!firebaseConfig?.functions || !firebaseConfig.functions?.source || !isString(firebaseConfig.functions.source)) {
		throw new Error('Error: Required "functions.source" field is missing from Firebase Configuration file.');
	}

	return {
		functions: {
			name: rewriteConfig.function ?? rewriteConfig.run.serviceId,
			source: path.join(path.dirname(firebaseJson), firebaseConfig.functions.source),
			runtime: firebaseConfig.functions?.runtime,
		},
		publicDir: path.join(path.dirname(firebaseJson), hostingConfig.public),
	};
}

/**
 * Cloud Run Service ID rules:
 * - only lowercase alphanumeric characters and dashes
 * - cannot begin or end with a dash
 * - cannot be longer than 63 characters
 * @param {string} serviceId
 * @returns {boolean} `true` if valid
 */
function validCloudRunServiceId(serviceId) {
	return /^[a-z\d][a-z\d-]+[a-z\d]$/gm.test(serviceId) && serviceId.length < 64;
}

/**
 * Cloud Function name rules:
 * - alphanumeric
 * - underscore
 * - max length 63 chars
 * @param {string} name
 * @returns {boolean} `true` if valid
 *
 * Rules a combination of
 * - https://github.com/firebase/firebase-tools/blob/1633f4fccbbc1bcbc6216fe13b8e888c8940bde4/src/deploy/functions/validate.ts#L38
 * - https://github.com/firebase/firebase-tools/blob/2dc7216a498dee2ca7e2acc33d6ba16d5647e27f/src/extractTriggers.js#L18
 */
function validCloudFunctionName(name) {
	return /^\w{1,63}$/.test(name);
}

/**
 * Ensure provided static asset output dir (firebase.json:hosting.public) is not the same as the source dir.
 * Throw error if invalid.
 *
 * @param {{
 * 	dest:string
 * 	source:string
 * }} param source and destination directory for static assets
 */
function ensureStaticResourceDirsDiffer({source, dest}) {
	if (source === dest) {
		throw new Error(`Error: "firebase.json:hosting.public" field (${dest}) must be a different directory to "svelte.config.js:kit.files.assets" field (${source}).`);
	}
}

/**
 * Validate Cloud Function runtime Node.js version for use with SvelteKit.
 * Throw error if invalid.
 * Returns number part of input values.
 *
 * @param {{
 * 	functionsPackageJsonEngine: undefined|string;
 * 	firebaseJsonFunctionsRuntime: undefined|string;
 * }} version
 * @returns {string} number part of input value. Eg: input=nodejs14 return=14 || input=14 return=14
 */
function ensureCompatibleCloudFunctionVersion({functionsPackageJsonEngine, firebaseJsonFunctionsRuntime}) {
	const validPackageJsonValues = [
		'14',
		'16',
		'18',
	];
	const validFirebaseJsonValues = [
		'nodejs14',
		'nodejs16',
		'nodejs18',
	];

	const validPkgJsonVersion = validPackageJsonValues.includes(functionsPackageJsonEngine);
	const validFirebaseJsonVersion = validFirebaseJsonValues.includes(firebaseJsonFunctionsRuntime);

	if (!validPkgJsonVersion && !validFirebaseJsonVersion) {
		throw new Error(`Error: Node.js runtime not supported. SvelteKit on Cloud Functions requires either "firebase.json:functions.runtime" with one of ${validFirebaseJsonValues} or "functions/package.json:engines.node" with one of ${validPackageJsonValues}`);
	}

	return functionsPackageJsonEngine
		? functionsPackageJsonEngine.slice(-2)
		: firebaseJsonFunctionsRuntime.slice(-2);
}

/**
 * Format message with relative dir on following newline.
 *
 * @param {string} message
 * @param {string} dir
 * @returns {string} formatted message with relative dir on following newline
 */
function logRelativeDir(message, dir) {
	return `${message}:\n\t${path.relative(process.cwd(), dir)}`;
}

export {
	ensureCompatibleCloudFunctionVersion,
	ensureStaticResourceDirsDiffer,
	logRelativeDir,
	parseFirebaseConfiguration,
	validCloudRunServiceId,
	validCloudFunctionName,
};
